components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java

```diff
@@ -72,7 +74,12 @@ public void handle(HttpServletRequest request, HttpServletResponse response,
 
         List<AuthenticatorConfig> authConfigList = stepConfig.getAuthenticatorList();
 
-        String authenticatorNames = FrameworkUtils.getAuthenticatorIdPMappingString(authConfigList);
+        String authenticatorNames;
+        try {
+            authenticatorNames = URLEncoder.encode(FrameworkUtils.getAuthenticatorIdPMappingString(authConfigList), "UTF-8");
```

Comment:
You must encode only at the point we build do the redirect by by passing the URL, not at the points we build the URL. Because later if someone adds a debug log to print the authenticatorNames without knowing you have encoded it the log will print a URL encoded value unnecessarily. Or else that person has to decode and print it which is unnecessary overhead. The best practice we have followed in the XSS hackathon is to encode it at the place we render the output.